wordpress.org/public_html/wp-content/plugins/wporg-profiles-wp-activity-notifier/wporg-profiles-wp-activity-notifier.php

```diff
@@ -51,6 +51,7 @@ private function __construct() {
 
 	public function init() {
 		add_action( 'transition_post_status',    array( $this, 'maybe_notify_new_published_post'   ), 10, 3 );
+		add_action( 'wp_insert_post',         	 array( $this, 'insert_post' ),                       10, 2 );
```

Comment:
I'm curious why you used the `wp_insert_post` action rather than `transition_post_status` (as used by `maybe_notify_new_published_post()`.

Reply:
I used `transition_post_status` in the beginning, but since the action is triggered before saving data to DB, I can't get the post meta with the function [get_post_meta](https://developer.wordpress.org/reference/functions/get_post_meta/).

Reply:
After thinking about it some more: you're on the right track, `wp_insert_post` is (almost) the right action to use here. If it was only on `transition_post_status` then we'd miss seeing presenters that were added to the post after it was published.

I think what's needed is actually the `wp_insert_post_data` filter. That's triggered during `wp_insert_post()`, but before the post is inserted or updated:

https://github.com/WordPress/wordpress-develop/blob/9b105d92a4b769f396ba798db1f106abab75001f/src/wp-includes/post.php#L4361-L4374

If you hook that one, the `$postarr` array should contain all the data being inserted, including the `meta_input` data that gets [inserted as postmeta](https://github.com/WordPress/wordpress-develop/blob/9b105d92a4b769f396ba798db1f106abab75001f/src/wp-includes/post.php#L4498). And the `$update` param should tell you if it's an update or an insert. So in the case of an insert, you'd add activity for all of those users. If it's an update, you could use `get_post_meta()` to figure out if any new presenters are being added, and only add activity for those.

Does that make sense? I think that would mostly deal with the issue of duplicates too, since you'd only notify when a new username is found in `$postarr`.

Reply:
Nice!
It would deal with the case: 
[run1] presenter: ren -> send activity to ren
[run2] presenter: ren, ren2 -> send activity to ren2

But if it's the following case, it still duplicates, but I believe it's acceptable.
[run1] presenter: ren, ren2 -> send activity to ren, ren2
[run2] presenter: ren -> send activity to no one
[run3] presenter: ren, ren2 -> send activity to ren2 again

Reply:
Yeah I agree - handling of that edge case isn't quite right, but it's good enough for now. 

If we also add some basic dupe handling on the API side then that should be pretty solid. Let's make that a separate issue. 

There's also the issue of removal, but since I don't think there's a delete API there's nothing we can do about it for now. We could consider handling that in a separate issue later if it becomes an issue.